Network ACLs + list_networks

kylehogan opened the pull request:
functionality to list all networks from #519 
depends on #545 as the access parameter for networks was changed

this closes #519, #545, #520, #453, and possibly #496 

the functionality for showing connected nodes is outside of show_network() currently, but could be implemented there instead

also affects #566, in a negative way

kylehogan commented on tests/unit/api.py:
```diff
@@ -1222,6 +1252,58 @@ def test_list_projects(self, db):
+        }
+
+    def test_list_network_attachments(self, db):
+        api.node_register('node-99', 'ipmihost', 'root', 'tapeworm')
```
any ideas on why this is failing the Travis checks? 
getting a type error that node_register takes only one argument
I am not getting the same error locally and none of the other tests are failing with this line

henn replied:
@SahilTikale - does the OBM IPMI change we just merged require a
different config file to load the correct drivers?

kylehogan replied:
it was a change to the node_register call that was required by the OBM IPMI change - I didnt have it locally so I didnt run into the issue there and all the other tests were getting fixed by git when it merged. 

I just changed my calls to node_register to match the new syntax and it was fine

henn commented on docs/rest_api.md:
```diff
@@ -88,6 +88,20 @@ Possible errors:
+
+`POST /project/<project>/add_network`
+
+Add a network to project named <project>.
```
Could you reflect that you are adding _access_ to a network for this and project_remove_network?

kylehogan commented:
@xuhang57 Lucas, would you mind reviewing this since the list_networks code is yours? I made a few small changes because of the many to many project:network relationship and that public networks should have project access be 'None', but everything else should be the same.

xuhang57 commented:
Thanks a lot Kyle! @kylehogan  And it looks good to me and I agree with you that we should set them as 'None', which you have already fixed.

kylehogan commented:
There is an issue with the ownership of admin created networks now that a network can be accessed by multiple projects. 

I believe that the original intent was that if an admin created a network with access 'project1' then project1 would effectively be the 'owner' of the network. 
Now that access is a list, all projects in the list have equal access rights to the network, but some functions such as adding a project to the network access list should only be allowed by an admin or the project that 'owns' the network.

If a network was created by a project then the owner is obviously network.creator, but for admin created networks I don't think restricting these functions to admin users is in line with what was originally intended by having an admin create a network to be accessed by a project.

I think it would make sense to have network specific admins to allow there to be a distinction between projects that are allowed to access the network and projects that are allowed to grant/remove access for other projects or delete the network etc.

kylehogan commented:
done pending review+discussion

gsilvis commented:
Going to put some comments from our conversation yesterday:

Prefer ``network_grant_access`` and ``network_revoke_access`` to the current ``project_add_network`` and ``project_remove_network``.

There are several auth subtleties that we need to A) decide B) document.  I'm going to put the questions, along with the answers I recall us generally agreeing on.

--Can the creator of a network remove other projects' nodes from it? [Yes]
--What information on other projects' attached nodes can the network creator see? [nic name, node name, channel; what's required for the remove call]
--What information on other projects' attached nodes can someone with access to a network see [none]
--If an administrator creates a network for a particular project, can that project delete it? [Uncertain---perhaps allow both methods?  Requires reworking 'create_network' a little bit.]
--When you revoke a project's access to a network, is it okay if they still have nodes on it? [I lean towards yes, but we definitely did not decide for sure.]

gsilvis commented on haas/test_common.py:
```diff
@@ -320,53 +320,66 @@ def initial_db():
             },
             {
-                # For some tests, we want things to initial be attached to a
+                # For some tests, we want things to initialyl be attached to a
```
there's still a typo here

gsilvis commented on haas/model.py:
```diff
@@ -274,7 +278,7 @@ def power_cycle(self):
         """
         assert False, "Subclasses MUST override the power_cycle method "
-
+    
```
extra whitespace

gsilvis commented on haas/api.py:
```diff
@@ -605,20 +721,29 @@ def show_network(network):
+    if network.access:
+        result['access'] = [p.label for p in network.access]
+    else:
+        result['access'] = None
```
Why ``None`` rather than ``[]``?

gsilvis commented on haas/api.py:
```diff
@@ -529,24 +645,24 @@ def network_create(network, creator, access, net_id):
-            raise BadArgumentError("Project-created networks must use network ID allocation")
-        access = _must_find(model.Project, access)
+            raise BadArgumentError("Project-owned networks must use network ID allocation")
+        access = [_must_find(model.Project, access)]    
```
extra whitespace

gsilvis commented:
Are we using ``[]`` or ``None`` for a network with public access?  I feel like I've seen both in this PR

kylehogan commented:
> Prefer network_grant_access and network_revoke_access to the current project_add_network and 
> project_remove_network.
Current names are project_grant/revoke_network_access

> --If an administrator creates a network for a particular project, can that project delete it? 
I went with 'no'. I changed the field 'creator' to 'owner' and if an admin creates a network for a project they would need to pass that project as the 'owner' parameter if they want that project to be the owner and thus be able to delete the network

> --When you revoke a project's access to a network, is it okay if they still have nodes on it?[=
I went with 'no' because then the access isn't really revoked. - we need to add functionality for the owner of a network to remove other projects nodes from it

> Are we using [] or None for a network with public access? I feel like I've seen both in this PR
the empty list is for internally in the code - JSON should pass back 'None' to be human readable 
(I think I found all the places 'None' was being used and changed it but if I missed some that should be fixed)
I do say 'if network.access is None' places because this will be true for empty lists

gsilvis commented:
Excellent, I like all of those answers well enough.  (We'll need @henn's approval too, of course).

One last question about 'owner' --- does the code currently allow admin to create a network owned by a normal project with a custom VLAN?  I recall that when I wrote the code, I manually forbade that combination...

kylehogan commented:
> does the code currently allow admin to create a network owned by a normal project with a custom VLAN? I recall that when I wrote the code, I manually forbade that combination...

I am not sure. From looking at the auth tests I had thought it was allowed though.

gsilvis commented:
Yeah, this'll have to change.  Currently the logic is:

Check what the creator is.  If it's admin, require admin privileges.  If not, require project privileges, and disallow custom vlans and custom access.

It'll have to change to:

If the owner is admin, require admin privileges.  If the vlan is custom, require admin privileges.  Otherwise, require either admin /or/ the owner's project privileges.

[And of course, the access will have to change to being a list instead of a single project]

gsilvis commented:
@henn While @kylehogan works on this stuff, can you independently review the changes?

henn commented:
I'm leaning towards making these calls network-centric instead of project, since fundamentally the thing that is being managed is the network and not the project.

What do you guys think?

So instead of `POST /project/<project>/grant_network_access`, something like:
`POST /network/<network>/access/<new_project>` and `DELETE /network/<network>/access/<project>`.

List would then be: `GET /network/<network>/access`

henn commented:
ahh.. I see that @gsilvis recorded that from our conversation :-)

So far we've (mostly) stuck with using objects in our API rather than being procedural. Thus, it'd be good to continue that trend.

gsilvis commented:
Good point---I actually forgot to look at the proposed REST api, and just jumped to the python api... The only suggestion I'd make is that it should be ``PUT /network/<network>/access/<new_project>`` and ``DELETE /network/<network>/access/<project>``

Also, I'm assuming neither of these work on a public network?  Certainly they don't do what you might expect them to do.

henn commented:
> Good point---I actually forgot to look at the proposed REST api, and just jumped to the python api... The only suggestion I'd make is that it should be PUT /network/<network>/access/<new_project> and DELETE /network/<network>/access/<project>
> 

I don't feel strongly on PUT vs POST, though the reason I chose POST was because we were modifying an existing network object.

http://stackoverflow.com/questions/630453/put-vs-post-in-rest

gsilvis commented:
This is an idempotent operation that creates the resource specified by the URL, and the resource can later be DELETEd, and that operation is also idempotent.  Definitely feels like a PUT to me.

knikolla commented:
Agree with @gsilvis, definitely feels like put.

kylehogan commented:
Errors I'm seeing are:

> InternalError: (psycopg2.InternalError) cannot drop table nic because other objects depend on it
DETAIL:  constraint networkattachment_nic_id_fkey on table networkattachment depends on table nic
constraint networkingaction_nic_id_fkey on table networkingaction depends on table nic

and, after a call to initial_db:

> in initial_db.filter_by(label='pub_default').one()
self = flask_sqlalchemy.BaseQuery object at 0x7f53aa652dd0
E   MultipleResultsFound: Multiple rows were found for one()

I am not sure where to start with these, particularly the first one. Only the migrations tests were failing as of c86dc19, then with the changes starting at 3daeb17 these two errors showed up and the majority of the tests for postgres are failing/erroring

kylehogan commented:
last error is that the 'alembic_version' tables do not match. 

> ["  {u'alembic_version': {'rows': [(u'099b939261c1',),",
          "                                 (u'6a8c19565060',),",
          "                                 (u'b5b31d19257d',),",
          "-                                (u'df8d9f423f2b',),",
          "+                                (u'df8d9f423f2b',)],",
          '?                                                  +\n',
          "-                                (u'f4a1f5d7ec43',)],",

Not sure how to fix this. @zenhack any ideas?

zenhack commented on haas/migrations/versions/f4a1f5d7ec43_network_acl.py:
```diff
@@ -0,0 +1,49 @@
+
+# revision identifiers, used by Alembic.
+revision = 'f4a1f5d7ec43'
+down_revision = None
```
Not certain that this is your problem, but down_revision should match the "Revises" line in the docstring. The migrations doc says to change it to None only for new extensions, not for haas core. Perhaps that difference should be called out explicitly.

zenhack commented on haas/migrations/versions/f4a1f5d7ec43_network_acl.py:
```diff
@@ -0,0 +1,49 @@
+
+
+def upgrade():
+    ### commands auto generated by Alembic - please adjust! ###
```
Probably best to remove these comments after writing the script.

kylehogan commented:
@zenhack same error still. and I agree that we should note this in the docs (I had a few commits further back that had the down_revision set correctly and changed it later after reading the docs)

henn commented:
> last error is that the 'alembic_version' tables do not match.
> 
>     [" {u'alembic_version': {'rows': [(u'099b939261c1',),",

@zenhack - do you think we could ignore the `alembic_version` element in the check, since that would be an expected difference? One DB was created new and one was migrated to that point and would thus have additional versions.

zenhack commented:
@henn, that is *not* an expected difference. Creation should stamp the db with the current version, and upgrading should change it. It should contain one entry for the latest version of each branch (one for haas core, and one for each extension that has migrations).

I'll need to stare at  the error a bit more

kylehogan commented:
@zenhack 

I pushed this because when I ran the migration again to see if anything changed, the same table had an error, but the format was slightly different

here are the commands I ran to generate the new migration file:

```
git checkout master
sudo -i -u postgres
dropdb haas_db
createdb -O haas_user haas_db (my db account is haas_user)
su haas-user (my user account is haas-user)
haas-admin db create
haas-admin db upgrade heads
git checkout list_networks
haas-admin db migrate --head haas@head -m "networkACL2"
```

I then put the contents of upgrade and downgrade from my previous migration file into this one so the other tables would migrate correctly. 

new error looks like this: 
> ["  {u'alembic_version': {'rows': [(u'099b939261c1',),",
          "-                                (u'89630e3872ec',),",
          '?                                      ^ ^^^^^^^\n',
          "+                                (u'6a8c19565060',),",
          '?                                   ++ ++ + ^ ^^\n',
          "                                 (u'b5b31d19257d',),",
          "                                 (u'df8d9f423f2b',)],",`

compared to old: 
> [" {u'alembic_version': {'rows': [(u'099b939261c1',),",
" (u'6a8c19565060',),",
" (u'b5b31d19257d',),",
"- (u'df8d9f423f2b',),",
"+ (u'df8d9f423f2b',)],",
'? +\n',
"- (u'f4a1f5d7ec43',)],",

For reference, my migrations are `89630e3872ec` and `f4a1f5d7ec43`

zenhack commented:
@kylehogan, I tracked down the issue. I went ahead and fixed it, submitted a pr to your repo.

kylehogan commented:
@zenhack thanks!

I put the name back to just network ACL

kylehogan commented:
@gsilvis can you review this again now that it isn't failing tests?

@zenhack would you mind being the second reviewer since you've looked at it a bit for the migration issue?

zenhack commented on docs/rest_api.md:
```diff
@@ -135,27 +135,23 @@ The result must contain the following fields:
-
-* "access" -- if this is present, it is the name of the project which
-  has access to the network. Otherwise, the network is public.
+* "access", a list of projects that have access to the network or 'None' if the network is public
```
Should probably say `null` instead of `'None'`, this is documenting JSON output (whereas `None` is a python value). Be careful about the quotes, since the string `"null"` is different than the JSON literal `null`.

kylehogan replied:
I believe that the correct value is actually the python None since this is what is returned from the call. At least in my tests for show_network asserting that the value is None works correctly.

zenhack replied:
Right, but this is the API docs, so we should be talking in terms of what ends up on the wire. Python `None` gets mapped to JSON `null`.

zenhack commented on docs/rest_api.md:
```diff
@@ -173,6 +169,92 @@ Additionally, the `show_networks` api call may return the channel identifier
+The response must contain the following fields:
+
+* "network", the name of a network
+* "driver_id", the id of the network
```
Can we rename this to `"network_id"`? that's what it's called everywhere else.

kylehogan replied:
it isn't actually the network_id field, it's the label 
the value for 'driver_id' is the network_id field 
this from a previous request by @xuhang57 so he should also comment on changes

zenhack commented on docs/rest_api.md:
```diff
@@ -173,6 +169,92 @@ Additionally, the `show_networks` api call may return the channel identifier
+
+List all networks.
+
+Returns a JSON dictionary of dictionaries, indexed by the name of the network with interior dictionaries indexed by project and channel.
```
1. Please hard-wrap this line to 72-80 chars.
2. What does it mean to be indexed by both project and channel? Keys are just strings, so it's not clear what this means.
3. Your example below does not agree with this description.

zenhack commented on docs/rest_api.md:
```diff
@@ -173,6 +169,92 @@ Additionally, the `show_networks` api call may return the channel identifier
+
+Returns a JSON dictionary of dictionaries, indexed by the name of the network with interior dictionaries indexed by project and channel.
+
+The response must contain the following fields:
```
Probably should say something like "each internal dictionary in the response."

zenhack commented on docs/rest_api.md:
```diff
@@ -173,6 +169,92 @@ Additionally, the `show_networks` api call may return the channel identifier
+
+Authorization requirements:
+
+Only admins can list the networks.
```
Should probably stick to the formulaic language used everywhere else in the document, i.e:

* Administrative access

It may not seem like a big deal, but with specifications you want to leave little room for interpretation. Using the same phrasing for the same thing helps with that. Not a huge problem in this case, but generall good practice.

zenhack commented on docs/rest_api.md:
```diff
@@ -173,6 +169,92 @@ Additionally, the `show_networks` api call may return the channel identifier
+
+List all nodes that are attached to network <network>.
+    
+If optional argument 'project' is supplied will only list attached nodes belonging to the specified project.
```
Where does this argument come from? There's no body in a GET request, is it a query parameter? I don't think our `rest_call` decorator can even do that (unless you've changed it). This needs clearer specification.

Also, please hard-wrap to 72-80 chars.

zenhack commented on docs/rest_api.md:
```diff
@@ -173,6 +169,92 @@ Additionally, the `show_networks` api call may return the channel identifier
+    
+If optional argument 'project' is supplied will only list attached nodes belonging to the specified project.
+
+Returns a JSON dictionary of dictionaries with first level key being the name of the attached node and second level keys being:
```
Hard-wrap

zenhack commented on docs/rest_api.md:
```diff
@@ -173,6 +169,92 @@ Additionally, the `show_networks` api call may return the channel identifier
+Remove <project> from access list for <network>.
+
+If the network or project does not exist a NotFoundError will be raised.
+If the project is the owner of the network a BlockedError will be raised.
```
Everywhere else in this document we talk about HTTP status codes; keep in mind the separation of python code from the HTTP representation. Granted, we do return the names in the body of the response, but this should be consistent with the rest of the document.

zenhack commented on docs/rest_api.md:
```diff
@@ -173,6 +169,92 @@ Additionally, the `show_networks` api call may return the channel identifier
+
+Authorization requirements:
+
+Admins or the project that is the owner can list all attached nodes. Other projects can only list their own nodes.
```
I'm less than thrilled with doing something different based on the access level of the caller; I'd rather have the caller specify which of these they want, and then be given access or not accordingly. Maybe two calls:

* `GET /network/<network>/attachments/<project>` returns attachments for nodes owned by `<project>`. Requires access to `<project>`, output looks like:

```
     {
         "node": <node-name>,
         "channel", <channel>,
         "nic": <nic-name>
     }
```

* `GET /network/<network>/attachments` gets all attachments. requires admin or owner access. Result is a dictionary mapping project names to values like the result from the above call.

kylehogan replied:
this is what it does actually. if an admin calls it they can choose to specify a project, but are not required to whereas if a project calls it the request will only be successful if they also specified a project and that project is the same as the calling project (or the calling project is the owner of the network in which case they can see all attachments as if they were admin)

I'm unfamiliar with the specifics of how `rest_call` works actually so if you had pointers for how this should look I would appreciate it. particularly as it would relate to the optional project parameter - which from my understanding should be a query parameter

zenhack replied:
The docstrings in `haas.rest` are fairly detailed. We don't currently provide a way to get at the query string, however. One option is to tweak the interface so the project comes out of the path. you'd need multiple functions decorated with `rest_call`, but they could call out to a common helper function.

@gsilvis, @henn, any thoughts on this?

henn replied:
Either way (single function that treats project as optional or multiple functions) works for me.

Does rest_call() support being instantiated multiple times, like `route` is in [this stackoverflow](https://stackoverflow.com/questions/14032066/flask-optional-url-parameters)? If that's the case, we could mark project as an optional parameter in the python function and have two URL patterns that lead to it. That stackoverflow also has other ways to do it too.

Otherwise, having 2 functions (like list_network_attachments and list_network_attachments_project) works too.

If `rest_call()` would need to be modified, then to keep this PR simple we should probably go with the second option.

zenhack replied:
@henn, I hadn't realized you could do that; it should work since we're just calling `add_url_rule` ourselves (which is basically a trivial non-decorator version of `route`). Give it a shot.

henn replied:
@kylehogan - what do you think?

henn replied:
It looks like the current code adds `project` as an optional parameter. Does that address your concerns, @zenhack?

@kylehogan - could you append something to the effect of: `by passing in a "project" parameter specifying the project whose nodes the caller would like to list`

zenhack commented on haas/api.py:
```diff
@@ -133,7 +133,60 @@ def project_detach_node(project, node):
+
+   if project in network.access:
+       raise DuplicateError('Network %s is already in project %s'%
+                            (network.label, project.label))
```
This error is not documented in `rest_api.md`.

zenhack commented on haas/api.py:
```diff
@@ -133,7 +133,60 @@ def project_detach_node(project, node):
 
+@rest_call('PUT', '/network/<network>/access/<project>')
+def network_grant_project_access(project, network):
+   """Add access to <network> to <project>.
```
Per PEP257, you should have a blank line after the one-line summary. Same goes for the other docstrings.

zenhack commented on haas/api.py:
```diff
@@ -133,7 +133,60 @@ def project_detach_node(project, node):
+    
+    if network.access:
+        for proj in network.access:
+            authorized = authorized or ((proj.label == project.label) and get_auth_backend().have_project_access(proj))
```
This line is very long, could you wrap it?

zenhack commented on haas/api.py:
```diff
@@ -133,7 +133,60 @@ def project_detach_node(project, node):
+            authorized = authorized or ((proj.label == project.label) and get_auth_backend().have_project_access(proj))
+            
+    if not authorized:
+        raise AuthorizationError("You do not have access to this project.")
```
s/project/network/

zenhack replied:
Err, that's not right either. Probably something like "You are not authorized to remove that project from this network."

zenhack commented on haas/api.py:
```diff
@@ -133,7 +133,60 @@ def project_detach_node(project, node):
+        raise AuthorizationError("You do not have access to this project.")
+   
+    if project not in network.access:
+        raise NotFoundError("Network %s is not in project %s"%
```
Suggest using `%r` instead of `%s`, which will quote the value (when it's a string anyway).

zenhack commented on haas/api.py:
```diff
@@ -133,7 +133,60 @@ def project_detach_node(project, node):
+                            (network.label, project.label))
+
+    if project is network.owner:
+        raise BlockedError("Project %s is owner of network %s and its access cannot be removed"%
```
Same comment about the `%s`.

zenhack commented on haas/api.py:
```diff
@@ -133,7 +133,60 @@ def project_detach_node(project, node):
+    num_attachments = 0
+    for attachment in network.attachments:
+        if attachment.nic.owner.project.label == project.label:
+            num_attachments += 1
```
Rather than counting them and then checking after, you could just raise the exception in this if statement.

zenhack commented on haas/api.py:
```diff
@@ -274,7 +327,7 @@ def _have_attachment(nic, query):
         raise BlockedError("A networking operation is already active on the nic.")
 
-    if (network.access is not None) and (network.access is not project):
+    if (network.access) and (project not in network.access):
```
Why not just `if project not in network.access`?

kylehogan replied:
It errors out

if I'm remembering right it's because network access is sometimes None. It might make sense to change the way  network access is handled for public networks.

zenhack replied:
Getting a fresh look at it I'm realizng they're not semantically equivalent anyway; without the first part it won't do the right thing with public networks. Nevermind.

zenhack commented on haas/api.py:
```diff
@@ -479,7 +532,7 @@ def headnode_connect_network(headnode, hnic, network):
     project = headnode.project
 
-    if (network.access is not None) and (network.access is not project):
+    if (network.access) and (project not in network.access):
```
Same comment about the condition

zenhack commented on haas/api.py:
```diff
@@ -505,21 +558,65 @@ def headnode_detach_network(headnode, hnic):
+    return json.dumps(result, sort_keys = True)
+
+@rest_call('GET', '/network/<network>/attachments')
+def list_network_attachments(network, project=None):
```
So, as written this will try to fish the project arg out of the body of the request, problem:  GET requests don't have bodies.

Did we test this? If it works it's because the libraries aren't being as strict about HTTP semantics as they should be. If not, we should have.

See also my comments in the api docs.

zenhack commented on haas/api.py:
```diff
@@ -505,21 +558,65 @@ def headnode_detach_network(headnode, hnic):
+            node = {
+                'nic': attachment.nic.label, 
+                'channel': attachment.channel, 
+                'project': attachment.nic.owner.project.label
```
Is owner not already a project?

zenhack commented on haas/api.py:
```diff
@@ -529,24 +626,24 @@ def network_create(network, creator, access, net_id):
-        if access != creator.label:
-            raise BadArgumentError("Project-created networks must be accessed only by that project.")
+        if access != owner.label:
+            raise BadArgumentError("Project-owned networks must be accessed only by that project.")
```
Not true anymore. This call needs tweaking for the changes to how access works. Probably access should be a list of projects that should (initially) have access.

kylehogan replied:
I wasn't sure how to handle passing in a single project or optionally a list so the way it currently works is that the initial access for project created networks must be that project then other projects can be added once the network is created

zenhack replied:
It's probably fine to leave it as for now, but the error message needs to change somehow. Perhaps "Project-owned networks must be accessible by the owner."

Fyi, to recieve non-string arguments, you'd need to explicitly specify a schema (these will be mandatory soon anyway). Read the comments in `haas.rest` and the docs for the schema library:

https://pypi.python.org/pypi/schema

zenhack commented on haas/api.py:
```diff
@@ -133,7 +133,60 @@ def project_detach_node(project, node):
+            num_attachments += 1
 
+    if num_attachments != 0:
+        raise BlockedError("Project still has nodes attached to the network")
```
We need to check for headnodes too I think.

kylehogan replied:
would those not be in network.attachments?

zenhack replied:
No, they wouldn't. The way attachments work with headnodes is a bit different, basically because we've been looking to rip them out of the codebase for a while, and due to the fact that you can just create more headnode nics, we didn't need the ability to attach one hnic to more than one network. There's an `hnics` attribute (declared as a backref in the `Hnic` class) that you can check. See also #598 :(

zenhack commented on haas/api.py:
```diff
@@ -605,20 +702,29 @@ def show_network(network):
-        auth_backend.require_project_access(network.access)
+    if network.access:
+        #auth tests: admin owned networks with a project access
+        #TODO only network owner allowed to show the network?
```
We should decide on an answer to this question and implement it before merging.

kylehogan replied:
we discussed briefly and ended up with the code as it is currently so it will stay that way unless someone objects
I left the TODO so reviewers know that they might want to object

zenhack replied:
The thing that bugs me is that this is inconsistent with the decisions we've made on some of the other calls; we don't allow non-owners to see what nodes other projects have on the network, but this allows them to see what projects are there. My inclination is to require ownership of the network for this, but I'd like to get other dev's opinions. @henn, @gsilvis?

Once we come to a decision (either way), the comment should be removed.

henn replied:
Anyone who can attach to a network needs to be able to run `show_network` in order to get the channels, so this comes down to what other information should be shown.

In terms of other projects with ACLs, I don't have a strong opinion though lean towards the current behavior. For example, if BU has a common provisioning network they let only BU people connect to, it might be ok for them to see one another.

Another aspect is that a user could use the list of other allowed attachers to determine how they treat the network. For example, if I look at the ACL and see only trusted entities, I might not lock down my nodes' interfaces as much as if I saw projects I wasn't familiar with.

For public networks (like the Internet), not showing other projects is reasonable since that would basically let you enumerate the projects.

zenhack replied:
I don't have a very strong opinion. As long as we come to a clear decision and document it I'm happy.

zenhack commented on haas/model.py:
```diff
@@ -49,13 +48,20 @@
 
+    If `create` is True, this will generate the schema for the database, and
+    perform initial population of tables.
+
```
Where did this come from? I think I deleted this phrasing as part of my migrations work, perhaps it snuck back in as part of a botched merge? The  `create` argument doesn't exist, so this is obviously wrong.

zenhack commented on haas/model.py:
```diff
@@ -49,13 +48,20 @@
         uri = cfg.get('database', 'uri')
     app.config.update(SQLALCHEMY_DATABASE_URI=uri)
 
+# A joining table for networks and projects, which have a many to many relationship:
```
Let's pick a name that makes it clear what the relationship here actually is -- projects' *access* to networks.

zenhack commented on haas/test_common.py:
```diff
@@ -127,6 +127,7 @@ def newDB():
 def releaseDB():
     """Do we need to do anything here to release resources?"""
     with app.app_context():
+        db.reflect()
```
Explain this one to me?

kylehogan replied:
weird thing I found on the internet. Have no idea what it does. Introduced in commit of the name "I have literally no idea why this worked". The comment right above it describes the problem I was having that it fixed (somehow)

zenhack replied:
I don't see a comment...?

What it does is try to infer the database schema by actually looking at the database, as opposed to assuming it reflects the models declared in the code. I don't want to just shove this in there without at least understanding why.

zenhack replied:
So, that kind of error can happen if the database and the running code disagree on what the schema (equivalently, set of models) should be. This very well might happen if the migration tests are failing; that's half of what they're there to test. I wouldn't be surprised if, having fixed those tests, this was no longer necessary. Furthermore, if this *is* failing it's probably an actual bug of some sort. Can you try taking it out and see if it still fails?

zenhack commented on haas/test_common.py:
```diff
@@ -286,6 +287,33 @@ def undefine_headnodes():
 
     request.addfinalizer(undefine_headnodes)
 
+def additional_db():
```
Would you add a docstring explaining what this function does/is for?

kylehogan replied:
Do you have any ideas on what would be the most clear way to explain it?

zenhack replied:
iirc, this is just adding more database objects needed by the new stuff that weren't already in `initial_db`, correct? I'd explain the situation, perhaps starting with a one liner saying it populates the databse with initial_db + objects for testing network acls, and then a paragraph or two explaining why it needs to be an add on function, rather than just being added to initial_db (migration tests). Might make sense to add some working to the docstring for `initial_db` mentioning that the set of objects it creates can't be changed and why.

zenhack commented on tests/unit/api/main.py:
```diff
@@ -132,6 +132,54 @@ def test_duplicate_project_create(self):
         with pytest.raises(api.DuplicateError):
             api.project_create('acme-corp')
 
+class TestProjectAddDeleteNetwork:
```
We should be moving away from this style of test, where we populate the objects needed by the test *in* the test; it makes it harder to see at a glance what the test is actually looking for. This is why I created `initial_db`. If you need additional objects to write these tests, you can add them to `additional_db`, but please don't add more tests like this. They were a mistake when we wrote them, let's not make it worse.

I wrote a blog post inspired by this file a while back:

https://zenhack.net/2016/03/26/test-code-is-still-code.html

kylehogan replied:
what is a good way to not write them this way? the auth tests don't do this, but with the current structure of main.py I'm not sure how to make my few tests look like that without rewriting the rest of them

zenhack replied:
Two options: (1) use `additional_db`. If you need some extra objects in there to make these tests work, add them. (2) add a `setUp` function to the class that creates the projects and a few networks that are in an initial state that you can just use.

General rule of thumb, try not to have API calls besides the one you're testing in the test. Also, actually state explicitly what property the test is checking

If you have tests with incompatible requirements for the database comments, try to split them into (large) groups of compatible requirements, and split them into classes based on that (with their own setup function).

Also, make sure you're explicitly stating what the tests are checking -- what about these api calls are you testing? (looks like checking that side effects of the calls actually take place).

zenhack commented:
Done with my first pass of comments.

zenhack commented:
Wanted to mention: don't be discouraged by the size of this batch of comments; the pr is pretty good. For the most part, it's just *big*, and number of problems per line of code tends to be pretty constant for a given codebase/set of developers.

Sorry for dumping a huge pile of work on you :\

kylehogan commented:
I think I got back to you with questions on most of your comments and the ones that I didn't have questions on I just addressed in my local version. Haven't pushed it yet because from my understanding github will squash your set of comments when I do and they'll be harder to read. Thanks!

henn commented on haas/api.py:
```diff
@@ -529,24 +626,24 @@ def network_create(network, creator, access, net_id):
-        creator = _must_find(model.Project, creator)
-        auth_backend.require_project_access(creator)
+    # Check authorization and legality of arguments, and find correct 'access' and 'owner'
+    if owner != "admin":
```
Should this be:
```python
if not auth_backend.have_admin():
...
```

henn commented on docs/networks.md:
```diff
@@ -3,14 +3,13 @@
-- 'access' represents which project's nodes can access it, or '' (the
-  empty string) if all   projects' nodes can.  This will eventually be extended
-  into a ACL-style approach.
+- 'access' represents which projects' nodes can access it, or [] (the
```
`[]` isn't rendering here. You can check it either by playing around with one of these comments boxes and hitting "Preview", or by clicking the document button in the diffs.

Maybe try: 
```markdown
`[]`
```

henn commented on docs/networks.md:
```diff
@@ -3,14 +3,13 @@
-- 'creator' represents what project created it; or 'admin' if it was
-  administrator-created.  If there are still networks created by a project,
-  then that project cannot be deleted.  Only the creator has the ability to
+- 'owner' represents what project owns it; or 'admin' if it was
```
If we're going to keep "admin" as a special project name, we should either a) stick a restriction in `project_create` and document or b) make it an illegal project name (like <admin>), though this would only apply once #600 is implemented.

zenhack replied:
Maybe pass in null instead of a string? The current interface dates back to when everything was form data, so that wasn't an option.

henn replied:
`null` sounds good. What do you think, @kylehogan ?

kylehogan replied:
I don't have an objection, but am not sure I understand the distinction here. I didn't think python recognized `null` so if it was inputting/outputting that it would either need to be `None` or the string `'null'`. Would it be handled in the same way that `None` is the access for public networks?

henn replied:
You can pass in `null` to json, and it will come through as `None` in python.

From `ipython`:

```python
In [10]: foo = {'a':None}

In [11]: foojson = json.dumps(foo)

In [12]: foojson
Out[12]: '{"a": null}'

In [13]: json.loads(foojson)
Out[13]: {u'a': None}
```

kylehogan commented:
@zenhack I did not make the recommended changes to the tests in the latest update. When I used `additional_db ` with the tests for main.py some of them failed - not because they did not have enough objects, but because additional_db contained more objects than they were expecting (I think). 

I do think that this should get changed, but possibly not in this pr since the failing tests that would need to be fixed didn't have anything to do with the network_acl

henn commented on haas/api.py:
```diff
@@ -505,21 +563,72 @@ def headnode_detach_network(headnode, hnic):
+            
+    return json.dumps(result, sort_keys = True)
+
+@rest_call('GET', '/network/<network>/attachments')
```
Please add a Schema{} here and in the other API calls. See the other API calls for examples of optional arguments.

After a soon-to-be-committed PR is merged, they will become mandatory.

henn commented on haas/api.py:
```diff
@@ -505,21 +563,72 @@ def headnode_detach_network(headnode, hnic):
+
+    if project is not None:
+        project = _must_find(model.Project, project)
+        for proj in network.access:
```
Instead of a for loop here, is there a reason not to use:
```python
if not (project in network.access and auth_backend.have_project_access(project)):
    raise ...
```

kylehogan replied:
I believe the reason that I did it this way is that the network owner is allowed to list nodes for a specific project as well - so they would be authorized, but auth_backend.have_project_access(project)) would be false.

henn replied:
Ahhh.

So maybe something like:
```python
if project is not None and not authorized:
    if not (project in network.access and auth_backend.have_project_access(project)):
        raise ...
```

henn replied:
This seems very similar to our discussion on line 177 (including that we need to check `network.owner` is not `None` first)

zenhack commented on docs/rest_api.md:
```diff
@@ -236,25 +240,32 @@ Admins or the project that is the owner can list all attached nodes. Other proje
+
+* 400, if:
+   * the network or project does not exist
+   * the project already has access to the network
```
Can you double check the status codes? I'm fairly certain the first is a 404, and I would guess the second is a 409.

zenhack commented on docs/rest_api.md:
```diff
@@ -236,25 +240,32 @@ Admins or the project that is the owner can list all attached nodes. Other proje
 
+Possible Errors:
+
+* 400 - If the network or project does not exist.
```
Again, would expect this to be a 404; check.

zenhack commented on haas/api.py:
```diff
@@ -560,21 +565,28 @@ def headnode_detach_network(headnode, hnic):
+    """Lists all the attachments from <project> for <network>
+
+    If <project> is not specified, lists all attachments for <network>
+    """
```
If you're not going to have anything else in the docstring, you should have the closing quote on the same line.

zenhack commented:
@kylehogan, I see that `test_list_networks` starts off by calling the api and verifying that it doesn't yield any results. adding additional_db is of course going to cause this to fail since it will add networks. What I'd do there is split it into (at least) two tests:

* A test that does no initial population, and just tests that the result is empty (on an empty db)
* Tests for the rest of the functionality, which can use what's in additional_db.

Another drawback of the current state of this file is we have a lot of tests that are actually testing more than one thing, which we should be avoiding.

Were there any other tests that were failing? If so, can they be fixed in a similar way?

kylehogan commented:
There were a few failing tests, I believe that they could be fixed similarly. Some of the problems stemmed from tests expecting the db to be empty, others expected a specific set of results. 

Is there a way to specify which database a specific test should use?

I was also wondering if it might be a problem that a lot of the tests would be attempting to add things to the database that already existed since additional_db doesn't start empty?

henn commented on haas/api.py:
```diff
@@ -133,7 +133,65 @@ def project_detach_node(project, node):
+    authorized = get_auth_backend().have_project_access(network.owner)
+    
+    if network.access:
+        for proj in network.access:
```
Could you replace this for loop with a conditional, similar to below:
```python
if not (project in network.access and auth_backend.have_project_access(project)):
    raise ...
```

kylehogan replied:
this is the same as the previous one you commented on in that the network owner is allowed to revoke the access of any project, but `auth_backend.have_project_access(project))` will be false.

I agree that this way of handling it is a bit messy, but I'm not sure how to cover the two separate cases - where the project is removing itself and where the network owner is removing a separate project - otherwise

henn replied:
What about this (with fixed up indentation and whatnot)?

```python
owner = network.owner

# must be admin, the owner of the network, or <project> to remove <project>. 
if not ( auth_backend().have_admin() or \
           (owner is not Null and auth_backend.have_project_access(owner)) or \
           (project in network.access and auth_backend.have_project_access(project)):
    raise ...
```

Note that `network.owner` can be `None`, so we need to check for that (ie - the `authorized` check above is not valid). As a side note, I just opened #623 to allow `None` to be passed, since that seems like a reasonable way to simplify the code.

henn commented on haas/api.py:
```diff
@@ -133,7 +133,65 @@ def project_detach_node(project, node):
+        raise BlockedError("Project %r is owner of network %r and its access cannot be removed"%
+                           (project.label, network.label))
+
+    for attachment in network.attachments:
```
Similar comment for this and the next for loop with `hnic`: Could we replace these with SQLAlchemy in order to maintain scalability? I think where ever possible, we should use the best tool for the job, which in cases like these is the underlying SQL

You can find some info on this here:
http://docs.sqlalchemy.org/en/rel_0_9/orm/tutorial.html#querying-with-joins

kylehogan replied:
Im getting an error:  `AttributeError: Neither 'InstrumentedAttribute' object not 'Comparator' object associated with NetworkAttachment.nic had an attribute 'owner'`

this is the query: model.NetworkAttachment.query.join(model.Project, model.NetworkAttachment.nic.owner.project.label == Project.label).count()

I am a) not quite sure what this error means and b) unsure of how to actually make a comparison between an network attachment and a project without going through the owner of the nic

henn replied:
@shwsun - could you help debug this?

henn replied:
Maybe this?

```python
db.session.query(NetworkAttachment).filter(NetworkAttachment.nic.owner.project == project)
.filter(NetworkAttachment.network == network).count()
```

Note - that would require a `from haas.model import NetworkAttachment`

kylehogan replied:
Same error.

zenhack commented:
@kylehogan, you should be able to apply fixtures to specific tests, and you can group them into classes and apply the fixture to the class to reduce the boilerplate. The pytest docs describe how to do this. This way you can make a few different fixtures similar to additonal/initial db, that create objects appropriate for different sets of tests.

henn commented on haas/api.py:
```diff
@@ -133,7 +133,65 @@ def project_detach_node(project, node):
     db.session.commit()
 
-
+@rest_call('PUT', '/network/<network>/access/<project>')
```
Could you add a `Schema` argument for this call and the next, similar to this:
```python
@rest_call('PUT', '/project/<project>', Schema({'project': basestring}))
def project_create(project):
```

Now that we've merged #596, all calls need a `Schema`.

henn commented on haas/api.py:
```diff
@@ -505,21 +563,73 @@ def headnode_detach_network(headnode, hnic):
                             # Network Code #
                             ################
 
+@rest_call('GET', '/networks')
```
Could you add a `Schema` argument, similar to this:
```python
@rest_call('GET', '/projects', Schema({}))
def list_projects():
```

Now that we've merged #596, all calls should have a `Schema`.

henn commented on haas/api.py:
```diff
@@ -505,21 +563,73 @@ def headnode_detach_network(headnode, hnic):
+            
+    return json.dumps(result, sort_keys = True)
+
+@rest_call('GET', '/network/<network>/attachments', schema=Schema({
```
Please add `network` to the Schema here.

kylehogan commented:
The two things that are not in this are the modifications to the tests requested by @zenhack and the switch from 'admin' to null for network owner as requested by @henn. I believe all other suggested changes have been made with the exception of my issue [here](https://github.com/CCI-MOC/haas/pull/554#discussion_r64279068)

I think that both of these should be separate pull requests (which I can get started once this is in)
Neither was introduced with the changes I made here and I would prefer that this pr doesn't get any bigger. (Also, it would be best if it wasn't still open when I left for Lincoln next week)

zenhack commented:
The thing that makes me uncomfortable is that by adding more of that kind of test we're exacerbating the situation; it's not just a matter of not fixing something that's already broken. I do appreciate that this PR has been something of an Odyssey.

I think leaving the loop in there is fine

kylehogan commented:
okay, redid some of the unit tests. @zenhack do they look okay?

zenhack commented on tests/unit/api/main.py:
```diff
@@ -1405,9 +1512,113 @@ def port_detach_nic_node_not_free(self):
+        api.project_connect_node('anvil-oldtimer', 'node-100')
+        api.network_grant_project_access('anvil-oldtimer', 'hammernet')
+        api.node_connect_network('node-100', '100-eth0', 'hammernet')
+        deferred.apply_networking()
```
This block probably should go in a fixture as well (or use objects in one of the existing fixtures if possible).

kylehogan replied:
added the new code but forgot to delete the old stuff... should be good now

zenhack commented:
@kylehogan, that looks a lot better. I made an inline comment on one test that could use the same treatment, but the rest of them look good.

zenhack commented:
I'm happy. +1

matsuuran commented:
@kylehogan everything looks good, but you need to add network to the Schema in api.py line 599

henn commented:
FYI - it would also be good to convert the for loops into SQLAlchemy statements before committing if possible.

@shwsun has offered to assist with this, as #621 needs this.

henn commented on haas/api.py:
```diff
@@ -138,7 +138,69 @@ def project_detach_node(project, node):
+        if attachment.nic.owner.project.label == project.label:
+            raise BlockedError("Project still has node(s) attached to the network")
+
+    for hnic in network.hnics:
```
Same comment as above

zenhack commented:
@henn, I don't follow, in what way is #621 dependant on the use of sqlalchemy expressions vs. loops?

While I agree we should not be using loops if there's a straightforward solution using sql, I see some futzing with getting a query right going on above, and this is looking like premature optimization. Until we've actually shown it to be too slow, we should be doing the simple thing, not the fast thing.

Also, this pr is already *enormous*, let's not add anything else if we don't have to.

shwsun commented:
@zenhack  #621 need the network ACL to be merged since some code is also from @kylehogan  and there are migration failures. I will be able to resolve the failures once this is done.

henn commented:
@zenhack thanks for your feedback!

Re: futzing, I think it's important to keep the basic design principle of letting the framework do this sort of stuff and want to keep the master branch in line with that. Otherwise, it'll take additional work (by people not potentially not as familiar with all of the nuances of how the code works) to convert this code when the scalability is needed. I also happen to think this code is a little more explicit and readable (3 lines for each of the 3 classes of users that can do things).

In addition, it led to noticing some bugs in the current way things are working, since network.owner can be `None`. BTW - I'd be interested in your thoughts on #623 to make it valid to pass `None` to the auth calls.

kylehogan commented on haas/api.py:
```diff
@@ -138,7 +138,68 @@ def project_detach_node(project, node):
+        raise NotFoundError("Network %r is not in project %r"%
+                            (network.label, project.label))
 
+    if project is network.owner:
```
project could be none and network.owner could be none (for admin owned network)

henn commented:
I think there may have been a copy & paste error in 69c19aeef4782292bfda253c6c566a08226395d0

I'd suggest backing it out.

henn commented:
@kylehogan I think there was a copy & paste error in 69c19aeef4782292bfda253c6c566a08226395d0

Suggest backing it out and trying again.

zenhack commented:
>    @zenhack #621 need the network ACL to be merged since some code
>    is also from @kylehogan and there are migration failures. I will be
>    able to resolve the failures once this is done.

Ah, I misunderstood what was being said; I thought @henn meant that the
loops were somehow a problem for that pr, which was very confusing. This
makes more sense.

zenhack commented:
So, actually, looking at that conditional, it's not obvious to me that we've even removed the loop. We've still got `project in network.access` in there. Do we know for certain that that actually shells out to the db instead of iterating?

shwsun commented:
@zenhack I agree that sentence is a little bit confusing. 

@henn @kylehogan and me have been looking into the loop problem for a while this afternoon. The ``for loop()`` is used to get rid of ``AttributeError while querying: Neither 'InstrumentedAttribute' object nor 'Comparator' has an attribute``. My thought was that either we need to add foreign key in order to resolve it or we have to use for loop like @kylehogan did. 
(I guess those few discussions are missing cause @henn  and @kyle did cherrypicking this afternoon)

matsuuran commented:
@shwsun @zenhack 
is this PR still moving or is it being done else where?

shwsun commented:
Kyle will give Jason push access, hopefully it will be done this sprint

On Monday, August 1, 2016, Nicholas Matsuura <notifications@github.com>
wrote:

> @shwsun <https://github.com/shwsun> @zenhack <https://github.com/zenhack>
> is this PR still moving or is it being done else where?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/CCI-MOC/hil/pull/554#issuecomment-236684128>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/APFwzcmrvhh0SFtnXH1SxRL_LXZ34R5Bks5qbksxgaJpZM4HxXyu>
> .
>


-- 

(Jethro) Shuwen Sun
Candidate of Master of Science

Graduate School of Arts & Sciences (GRS)

Boston University

zenhack commented:
Quoting (Jethro) Shuwen Sun (2016-08-01 15:49:49)

>    Kyle will give Jason push access, hopefully it will be done this sprint
>    On Monday, August 1, 2016, Nicholas Matsuura <notifications@github.com>
>    wrote:

Yeah, I don't think anyone's picked it up since Kyle left; my memory was
it was pretty close, but then our submitter left... If we've actually
assigned someone to pick it up hopefully it will get done.

shwsun commented:
Yes, Kyle has almost finished this one, it is why we don't want to assign another one to start a new PR. @henn will fix some small errors that he believe is not negligible and it will be ready to ship.

matsuuran commented:
@henn any update on the small errors?

shwsun commented:
@matsuuran I am working on it. We should kill it soon.

shwsun commented:
@henn  carefully remind you that your hil is also outdated, don't know whether or not this could help solve the branch conflict yet.

shwsun commented:
@henn I dont think the last 2 commit should be here, we have to get rid of it after all.

zenhack commented on haas/api.py:
```diff
@@ -534,26 +596,80 @@ def headnode_detach_network(headnode, hnic):
+    if project is None:
+        # List all nodes connected to network.
+        # Only the project that owns the network should be able to list attached nodes
+        auth_backend.require_project_access(network.owner)
```
I believe network.owner can be `None`, which denotes admin ownership

henn replied:
Good point!

Originally, I was expecting #636 to land first, which would have allowed `None`. Now that we aren't doing that, I'll have fix this.

zenhack commented on haas/api.py:
```diff
@@ -534,26 +596,80 @@ def headnode_detach_network(headnode, hnic):
+        # List all nodes connected to network.
+        # Only the project that owns the network should be able to list attached nodes
+        auth_backend.require_project_access(network.owner)
+    else:
```
I think right now this else is dead code; we'd talked about adding two routes that map to this api call (one with the project and one without) but I don't see the other route anywhere. That probably still needs to be added.

henn commented:
@shwsun - you're right. I got rid of the 2 travis CI commits.

henn commented:
Note that the code for passing GET arguments from the CLI, a13d140c076f7f1dc4e27c2ff73a0ee6cc8ed674, should be code reviewed under #643.

If changes happen there, I'll bring them over to this branch too.

henn commented:
Also, for functionality of the optional GET argument in list_network_attachments(), this depends on #644.

gsilvis commented:
This is listed as waiting on +2, but it has merge conflicts.

gsilvis commented:
Also, it has /sixty-eight/ commits, eight of which are internal merge commits, and quite a few of which are typo fixes that shouldn't have their own commits.

zenhack commented:
Indeed, some rebasing/squashing would not go amiss.

henn commented:
FYI - since some merging is necessary, the plan discussed between
myself, @pjd-nu and @shwsun is to close this over the next few days and
create a new PR that has this branch merged with the latest master, with
some portion of the commits squashed.
>
> Indeed, some rebasing/squashing would not go amiss.
>

zenhack commented:
@henn, sounds good to me.

naved001 commented:
Hi @henn is this pull request still active, or will a new pr be created with this branch merged? Thanks!

zenhack commented:
Spoke to @henn about the next things for me to work on, first thing on the list is getting this sorted out. I've got a branch on which I'm working on resolving the conflicts and cleaning up the history; I'll open a new pr once it's worth a look. I'm going to close this one.
